037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+Having “continuous” auto-rebalancing is out of scope for this proposal, the field is added for allowing a future implementation.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+```
+
+Finally, the user should be able to define what are the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) which are normally described via a `KafkaRebalance` custom resource.
+In any case, for rebalancing the cluster, an actual `KafkaRebalance` custom resource has to be automatically created by the cluster operator.
+
+### New `KafkaRebalanceTemplate` custom resource
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the rebalancing configuration he wants but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance`.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, it’s about exporting the rebalancing configuration out of the Kafka custom resource avoiding to pollute that resource and increase the size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available 
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify goals as well as all the other options like he already does today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration for the `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), to specify if auto-approval is desired or not when auto-rebalancing is enabled.
+The template to be used is referenced by a new `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the reference is missing, the cluster operator can create an “empty” `KafkaRebalance` custom resource (which means the optimization proposal will be based on Cruise Control default configuration). It will behave as the rejected alternative.
```

Comment:
Do we really want to default to CC's default configuration, or would it be better to fail the reconciliation? Consider that it might be deleted by mistake, so you'd pick up the default behaviour by accidentally some future time.

Reply:
Also IDK what "It will behave as the rejected alternative." means.

Reply:
> Do we really want to default to CC's default configuration, or would it be better to fail the reconciliation? Consider that it might be deleted by mistake, so you'd pick up the default behaviour by accidentally some future time.

I thought that the user could have some "good to go" configuration as defaults in Cruise Control to avoid filling the template each time. Of course, making the template mandatory put some guardrails on how to use this feature. I have no strong opinion against it. The `template` field would be mandatory and the operator has to check that the corresponding resource exists or fails the reconcile.

> Also IDK what "It will behave as the rejected alternative." means.

The rejected alternative works the same when the `autoRebalance` doesn't contain any rebalancing options so it creates an empty `KafkaRebalance` (which is the same when the template is not specified in the proposed solution).